### PR TITLE
refactor worker rebalancing to fix draining

### PR DESCRIPTION
**Pull request by vito**
previously, if a worker timed out draining and came back while still in
landing/retiring state, it would remain in that state forever, so that
rebalancing wouldn't accidentally switch it to 'running'

instead, this commit prevents rebalancing during draining. this is done
by merging the drain runner into the beacon and synchronizing so that
draining can't happen with a concurrent rebalance registration
in-flight.

### CODE REVIEW

**Comment by vito** [comment]
@cirocosta @xtreme-sameer-vohra this should fix the behavior we observed on the last Wings deploy which left workers in 'landing/landed' state

**Comment by vito** [comment]
@cirocosta :+1:

File: worker/beacon.go [hunk 1]
```diff
@@ -91,44 +103,110 @@ func (beacon *Beacon) Run(signals <-chan os.Signal, ready chan<- struct{}) error
...
+			atomic.StoreInt32(&beacon.drained, 1)
+
+			logger := beacon.Logger.Session("drain")
+
+			logger.Debug("received-drain-signal", lager.Data{
+				"signal": sig.String(),
+			})
+
+			// prevent rebalancing from switching the worker back to 'running'
+			rebalanceCh = nil
```
[the remark this answers is not in the document]
Comment (xtreme-sameer-vohra) [reply worker/beacon.go]: Yep, the default value is nil

**Comment by xtreme-sameer-vohra** [comment]
Late to the game but looked at it briefly and it looks good 👍

**Pull request merged**